lib/grpc/client/adapters/mint/connection_process/connection_process.ex

```diff
@@ -0,0 +1,395 @@
+  def handle_call({:stream_body, request_ref, body}, from, state) do
+    queue = :queue.in({request_ref, body, from}, state.request_stream_queue)
+
+    {:noreply, State.update_request_stream_queue(state, queue),
+     {:continue, :process_request_stream_queue}}
+  end
+
+  def handle_call({:cancel_request, request_ref}, _from, state) do
+    state = process_response({:done, request_ref}, state)
+
+    case Mint.HTTP2.cancel_request(state.conn, request_ref) do
+      {:ok, conn} -> {:reply, :ok, State.update_conn(state, conn)}
+      {:error, conn, error} -> {:reply, {:error, error}, State.update_conn(state, conn)}
+    end
+  end
+
+  @impl true
+  def handle_info(message, state) do
+    case Mint.HTTP.stream(state.conn, message) do
+      :unknown ->
+        Logger.debug(fn -> "Received unknown message: " <> inspect(message) end)
+        {:noreply, state}
+
+      {:ok, conn, responses} ->
+        state = State.update_conn(state, conn)
+
+        state =
+          if state.requests == %{},
+            do: state,
+            else: Enum.reduce(responses, state, &process_response/2)
+
+        check_connection_status(state)
+
+      {:error, conn, _error, _responses} ->
+        state = State.update_conn(state, conn)
+        check_connection_status(state)
+    end
+  end
+
+  @impl true
+  def handle_continue(:process_request_stream_queue, state) do
+    {{:value, request}, queue} = :queue.out(state.request_stream_queue)
+    {ref, body, _from} = request
+    window_size = get_window_size(state.conn, ref)
+    body_size = IO.iodata_length(body)
+    dequeued_state = State.update_request_stream_queue(state, queue)
+
+    cond do
+      # Do nothing, wait for server (on stream/2) to give us more window size
+      window_size == 0 -> {:noreply, state}
+      body_size > window_size -> chunk_body_and_enqueue_rest(request, dequeued_state)
+      true -> stream_body_and_reply(request, dequeued_state)
+    end
+  end
+
+  @impl true
+  def terminate(_reason, _state) do
+    :normal
+  end
+
+  defp process_response({:status, request_ref, status}, state) do
+    State.update_response_status(state, request_ref, status)
+  end
+
+  defp process_response({:headers, request_ref, headers}, state) do
+    if State.empty_headers?(state, request_ref) do
+      new_state = State.update_response_headers(state, request_ref, headers)
+
+      :ok =
+        new_state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:headers, headers)
+
+      new_state
+    else
+      :ok =
+        state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:trailers, headers)
+
+      state
+    end
+  end
+
+  defp process_response({:data, request_ref, new_data}, state) do
+    :ok =
+      state
+      |> State.stream_response_pid(request_ref)
+      |> StreamResponseProcess.consume(:data, new_data)
+
+    state
+  end
+
+  defp process_response({:done, request_ref}, state) do
+    :ok =
```

Comment:
A non-blocking aside: I wonder if all of these hard-matches might cause one request in the connection to kill all other requests if something goes wrong. Really depends on what the pipe functions do.

Any insights on this @beligante?

Reply:
I could - if that's okay with out - remove these and leave the telemetry stuff only for the response process.  This way the connection process will simply be responsible for distribute the data and ignore errors related to process it

Reply:
I'd rather have them explicitly match if we ensure this function only returns `:ok`
If errors should be ignored, they shouldn't be returned.
If this is a refactor, we can leave a to-do for a future PR

Reply:
Gotcha! Then we're good here. This will always return `:ok`